internal/s3client/s3client.go

```diff
@@ -0,0 +1,371 @@
+		// If the connection got closed (firwall, proxy, etc.)
+		// we should also retry, just like if we'd had a 500.
+		if err == io.ErrUnexpectedEOF {
+			return "", true, err
+		}
+
+		return "", false, err
+	}
+	if resp.StatusCode != 200 {
+		return "", false, NewRespError(resp)
+	}
+
+	// S3 will return an error under a 200 as well. Instead of the
+	// CompleteMultipartUploadResult that we expect below, we might be
+	// getting an Error, e.g. with InternalError under it. We should behave
+	// in that case as though we received a 500 and try again.
+
+	var r struct {
+		ETag string
+		Code string
+	}
+
+	err = xml.NewDecoder(resp.Body).Decode(&r)
+	closeErr := resp.Body.Close()
+	if err != nil {
+		// The decoder unfortunately returns string error
+		// instead of specific errors.
+		if err.Error() == "unexpected EOF" {
+			return "", true, err
+		}
+
+		return "", false, err
+	}
+	if closeErr != nil {
+		return "", true, closeErr
+	}
+
+	// This is what S3 returns instead of a 500 when we should try
+	// to complete the multipart upload again
+	if r.Code == "InternalError" {
+		return "", true, errors.New("S3 internal error")
+	}
+	// Some other generic error
+	if r.Code != "" {
+		return "", false, fmt.Errorf("CompleteMultipartUpload error: %s", r.Code)
+	}
+
+	return strings.Trim(r.ETag, "\""), false, nil
+}
+
+// AbortMultipartUpload aborts a multipart upload, discarding any
+// partly-uploaded contents.
+func (c *Client) AbortMultipartUpload(uploadID string) error {
+	v := url.Values{}
+	v.Set("uploadId", uploadID)
+	s := c.url.String() + "?" + v.Encode()
+	resp, err := c.retryRequest("DELETE", s, nil, nil)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 204 {
+		return NewRespError(resp)
+	}
+	_ = resp.Body.Close()
+
+	return nil
+}
+
+// PutMD5 attempts to write an md5 file in a ".md5" subdirectory of
+// the directory where the blob is stored, with retries. For example,
+// the md5 for blob https://mybucket.s3.amazonaws.com/gof3r will be
+// stored in https://mybucket.s3.amazonaws.com/.md5/gof3r.md5.
+func (c *Client) PutMD5(url *url.URL, md5 string) error {
+	var err error
+	for i := 0; i < c.nTry; i++ {
+		err = c.putMD5(url, md5)
+		if err == nil {
+			break
+		}
+	}
+	return err
+}
+
+// putMD5 makes one attempt to write an md5 file in a ".md5"
+// subdirectory of the directory where the blob is stored; e.g., the
+// md5 for blob https://mybucket.s3.amazonaws.com/gof3r will be stored
+// in https://mybucket.s3.amazonaws.com/.md5/gof3r.md5.
+func (c *Client) putMD5(url *url.URL, md5 string) error {
+	md5Reader := strings.NewReader(md5)
+
+	r, err := http.NewRequest("PUT", url.String(), md5Reader)
+	if err != nil {
+		return err
+	}
+	c.signer.Sign(r)
+	resp, err := c.httpClient.Do(r)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return NewRespError(resp)
+	}
+	if err := resp.Body.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var err500 = errors.New("received 500 from server")
```

Comment:
I realize this is merely a code transplant, but this seems like a good candidate for `const`?